File: backend-springboot/src/main/java/com/anime/backend/service/AnimeImportService.java
package com.anime.backend.service;

import com.anime.backend.model.Anime;
import com.anime.backend.repository.AnimeRepository;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestTemplate;

import org.w3c.dom.Document;
import org.w3c.dom.NodeList;
import org.w3c.dom.Element;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.io.ByteArrayInputStream;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
public class AnimeImportService {

    @Autowired
    private AnimeRepository animeRepository;

    @Autowired
    private ImportState importState;

    private final RestTemplate restTemplate = new RestTemplate();
    private final ObjectMapper objectMapper = new ObjectMapper();

    private static final Map<String, String> MAL_STATUS_MAP = new HashMap<>();
    static {
        MAL_STATUS_MAP.put("watching", "Watching");
        MAL_STATUS_MAP.put("completed", "Completed");
        MAL_STATUS_MAP.put("on-hold", "On-Hold");
        MAL_STATUS_MAP.put("dropped", "Dropped");
        MAL_STATUS_MAP.put("plan to watch", "Plan to watch");
        MAL_STATUS_MAP.put("plantowatch", "Plan to watch");
    }

    @Async
    public void startProcessing(String fileContent) {
        importState.reset();
        int added = 0, updated = 0, errors = 0;

        String content = fileContent.trim();
        if (content.startsWith("\uFEFF")) {
            content = content.substring(1).trim();
        }

        // 1. Try XML
        if (content.startsWith("<?xml") || content.startsWith("<myanimelist>")) {
            try {
                DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
                DocumentBuilder builder = factory.newDocumentBuilder();
                Document doc = builder.parse(new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8)));
                
                NodeList entries = doc.getElementsByTagName("anime");
                int total = entries.getLength();
                importState.setTotal(total);
                importState.logMessage("📄 Found " + total + " anime entries in MAL XML export.");

                for (int i = 0; i < total; i++) {
                    Element animeElem = (Element) entries.item(i);
                    String malIdStr = getTagValue("series_animedb_id", animeElem);
                    String title = getTagValue("series_title", animeElem);
                    String status = getTagValue("my_status", animeElem);
                    String scoreStr = getTagValue("my_score", animeElem);

                    if (malIdStr != null && title != null && !title.trim().isEmpty()) {
                        try {
                            int malId = Integer.parseInt(malIdStr);
                            String rawStatus = status != null ? status.trim().toLowerCase() : "";
                            String category = MAL_STATUS_MAP.getOrDefault(rawStatus, status != null ? status : "Uncategorized");
                            String url = "https://myanimelist.net/anime/" + malId;

                            Double personalRating = null;
                            try {
                                if (scoreStr != null && !scoreStr.trim().equals("0")) {
                                    personalRating = Double.parseDouble(scoreStr.trim());
                                }
                            } catch (Exception ignored) {}

                            boolean created = upsertAnime(malId, title.trim(), url, category, personalRating);
                            if (created) {
                                added++;
                                importState.logMessage("[" + (i + 1) + "/" + total + "] ✅ Added — " + title);
                            } else {
                                updated++;
                                importState.logMessage("[" + (i + 1) + "/" + total + "] 🔄 Updated — " + title);
                            }
                        } catch (Exception e) {
                            errors++;
                            importState.logMessage("[" + (i + 1) + "/" + total + "] ❌ Error: " + e.getMessage());
                        }
                    }
                    importState.setCurrent(i + 1);
                }
                finishImport(added, updated, errors);
                return;
            } catch (Exception e) {
                importState.logMessage("❌ XML parse error: " + e.getMessage());
                importState.setStatus("error");
                return;
            }
        }

        // 2. Try JSON
        try {
            Map<String, List<Map<String, Object>>> data = objectMapper.readValue(
                content, 
                new TypeReference<Map<String, List<Map<String, Object>>>>() {}
            );
            
            int total = data.values().stream().mapToInt(List::size).sum();
            importState.setTotal(total);
            importState.logMessage("📄 Found " + total + " anime entries in JSON format.");

            int current = 0;
            for (Map.Entry<String, List<Map<String, Object>>> entry : data.entrySet()) {
                String category = entry.getKey();
                for (Map<String, Object> item : entry.getValue()) {
                    Object malIdObj = item.get("mal_id");
                    Object titleObj = item.get("name");
                    Object urlObj = item.get("link");

                    if (malIdObj != null && titleObj != null) {
                        try {
                            int malId = ((Number) malIdObj).intValue();
                            String title = titleObj.toString();
                            String url = urlObj != null ? urlObj.toString() : null;

                            boolean created = upsertAnime(malId, title, url, category, null);
                            if (created) {
                                added++;
                                importState.logMessage("[" + (current + 1) + "/" + total + "] ✅ Added — " + title);
                            } else {
                                updated++;
                                importState.logMessage("[" + (current + 1) + "/" + total + "] 🔄 Updated — " + title);
                            }
                        } catch (Exception e) {
                            errors++;
                        }
                    }
                    current++;
                    importState.setCurrent(current);
                }
            }
            finishImport(added, updated, errors);
            return;
        } catch (Exception ignored) { }

        // 3. Try TXT
        String[] lines = content.split("\\r?\\n");
        List<TxtEntry> txtEntries = new ArrayList<>();
        String currentCat = "Uncategorized";

        Pattern regex = Pattern.compile("/anime/(\\d+)");

        for (String line : lines) {
            line = line.trim();
            if (line.isEmpty()) continue;
            if (line.startsWith("#")) {
                currentCat = line.substring(1).trim();
                continue;
            }
            if (line.contains("|")) {
                String[] parts = line.split("\\|", 2);
                String title = parts[0].trim();
                String url = parts[1].trim();
                Matcher matcher = regex.matcher(url);
                if (matcher.find()) {
                    txtEntries.add(new TxtEntry(Integer.parseInt(matcher.group(1)), title, url, currentCat));
                }
            }
        }

        int total = txtEntries.size();
        importState.setTotal(total);
        importState.logMessage("📄 Found " + total + " anime entries in TXT format.");

        for (int i = 0; i < total; i++) {
            TxtEntry entry = txtEntries.get(i);
            try {
                boolean created = upsertAnime(entry.malId, entry.title, entry.url, entry.category, null);
                if (created) {
                    added++;
                    importState.logMessage("[" + (i + 1) + "/" + total + "] ✅ Added — " + entry.title);
                } else {
                    updated++;
                    importState.logMessage("[" + (i + 1) + "/" + total + "] 🔄 Updated — " + entry.title);
                }
            } catch (Exception e) {
                errors++;
            }
            importState.setCurrent(i + 1);
        }

        finishImport(added, updated, errors);
    }

    private void finishImport(int added, int updated, int errors) {
        importState.logMessage("");
        importState.logMessage("✔️  Import complete! Added: " + added + " | Updated: " + updated + " | Errors: " + errors);
        importState.setStatus("done");
    }

    private String getTagValue(String tag, Element element) {
        NodeList nlList = element.getElementsByTagName(tag);
        if (nlList != null && nlList.getLength() > 0) {
            NodeList subList = nlList.item(0).getChildNodes();
            if (subList != null && subList.getLength() > 0) {
                return subList.item(0).getNodeValue();
            }
        }
        return null;
    }

    private static class TxtEntry {
        int malId; String title; String url; String category;
        TxtEntry(int m, String t, String u, String c) { malId=m; title=t; url=u; category=c; }
    }

    @Transactional
    public boolean upsertAnime(int malId, String title, String url, String currentCategory, Double personalRating) {
        Optional<Anime> existingOpt = animeRepository.findByMalId(malId);
        Anime anime;
        boolean created = false;

        if (existingOpt.isPresent()) {
            anime = existingOpt.get();
            anime.setCategory(currentCategory);
            if (personalRating != null) anime.setPersonalRating(personalRating);
            animeRepository.save(anime);
            return false; // Not created, just updated DB
        } else {
            anime = new Anime();
            anime.setMalId(malId);
            anime.setTitle(title);
            anime.setUrl(url);
            anime.setCategory(currentCategory);
            anime.setPersonalRating(personalRating);
            animeRepository.save(anime);
            created = true;
        }

        // Fetch from external APIs
        try {
            Thread.sleep(1500); // Rate limiting
            ResponseEntity<String> response = restTemplate.getForEntity("https://api.jikan.moe/v4/anime/" + malId, String.class);
            
            String imageUrl = null;
            String synopsis = null;

            if (response.getStatusCode().is2xxSuccessful()) {
                JsonNode root = objectMapper.readTree(response.getBody());
                JsonNode data = root.path("data");

                JsonNode images = data.path("images");
                if (images.has("webp") && images.path("webp").has("large_image_url")) {
                    imageUrl = images.path("webp").path("large_image_url").asText();
                } else if (images.has("jpg") && images.path("jpg").has("large_image_url")) {
                    imageUrl = images.path("jpg").path("large_image_url").asText();
                }

                if (data.has("score") && !data.path("score").isNull()) anime.setScore(data.path("score").asDouble());
                if (data.has("year") && !data.path("year").isNull()) anime.setYear(data.path("year").asInt());
                if (data.has("type") && !data.path("type").isNull()) anime.setType(data.path("type").asText());
                if (data.has("synopsis") && !data.path("synopsis").isNull()) {
                    synopsis = data.path("synopsis").asText();
                    anime.setSynopsis(synopsis);
                }
                
                if (data.has("title_english") && !data.path("title_english").isNull()) {
                    anime.setTitle(data.path("title_english").asText());
                }

            } else {
                String[] fallback = fetchKitsuFallback(title);
                imageUrl = fallback[0];
                anime.setSynopsis(fallback[1]);
                if (response.getStatusCode().value() == 429) {
                    Thread.sleep(10000);
                }
            }

            anime.setImageUrl(imageUrl);
            anime.setImageBase64(downloadImageAsBase64(imageUrl));
            animeRepository.save(anime);
        } catch (Exception ignored) {
            // Background fetch error ignored
        }

        return created;
    }

    private String[] fetchKitsuFallback(String title) {
        try {
            org.springframework.http.HttpHeaders headers = new org.springframework.http.HttpHeaders();
            headers.set("Accept", "application/vnd.api+json");
            org.springframework.http.HttpEntity<String> entity = new org.springframework.http.HttpEntity<>(headers);

            ResponseEntity<String> response = restTemplate.exchange(
                "https://kitsu.io/api/edge/anime?filter[text]=" + java.net.URLEncoder.encode(title, "UTF-8") + "&page[limit]=1",
                org.springframework.http.HttpMethod.GET,
                entity,
                String.class
            );

            if (response.getStatusCode().is2xxSuccessful()) {
                JsonNode root = objectMapper.readTree(response.getBody());
                JsonNode data = root.path("data");
                if (data.isArray() && data.size() > 0) {
                    JsonNode attrs = data.get(0).path("attributes");
                    String img = null;
                    if (attrs.has("posterImage")) {
                        JsonNode poster = attrs.path("posterImage");
                        if (poster.has("large")) img = poster.path("large").asText();
                        else if (poster.has("medium")) img = poster.path("medium").asText();
                        else if (poster.has("original")) img = poster.path("original").asText();
                    }
                    String syn = attrs.has("synopsis") ? attrs.path("synopsis").asText() : null;
                    return new String[]{img, syn};
                }
            }
        } catch (Exception ignored) { }
        return new String[]{null, null};
    }

    public String downloadImageAsBase64(String url) {
        if (url == null || url.isEmpty()) return null;
        try {
            ResponseEntity<byte[]> response = restTemplate.getForEntity(url, byte[].class);
            if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
                String base64 = Base64.getEncoder().encodeToString(response.getBody());
                String contentType = response.getHeaders().getContentType() != null ? 
                    response.getHeaders().getContentType().toString() : "image/jpeg";
                return "data:" + contentType + ";base64," + base64;
            }
        } catch (Exception ignored) { }
        return null;
    }
}

File: backend-dotnetWebApi/Services/AnimeImportService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Xml.Linq;
using backend_dotnetWebApi.Data;
using backend_dotnetWebApi.Models;
using Microsoft.Extensions.DependencyInjection;

namespace backend_dotnetWebApi.Services
{
    public class AnimeImportService
    {
        private readonly ImportState _importState;
        private readonly IServiceScopeFactory _scopeFactory;
        private readonly HttpClient _httpClient;

        private static readonly Dictionary<string, string> MalStatusMap = new(StringComparer.OrdinalIgnoreCase)
        {
            {"watching", "Watching"},
            {"completed", "Completed"},
            {"on-hold", "On-Hold"},
            {"dropped", "Dropped"},
            {"plan to watch", "Plan to watch"},
            {"plantowatch", "Plan to watch"}
        };

        public AnimeImportService(ImportState importState, IServiceScopeFactory scopeFactory)
        {
            _importState = importState;
            _scopeFactory = scopeFactory;
            _httpClient = new HttpClient();
        }

        public void StartProcessing(string fileContent)
        {
            Task.Run(async () => await ProcessUploadedFileAsync(fileContent));
        }

        private async Task ProcessUploadedFileAsync(string fileContent)
        {
            _importState.Reset();
            int added = 0, updated = 0, errors = 0;

            string content = fileContent.Trim();
            if (content.StartsWith("\uFEFF"))
            {
                content = content.Substring(1).Trim();
            }

            // 1. Try XML
            if (content.StartsWith("<?xml") || content.StartsWith("<myanimelist>"))
            {
                try
                {
                    var doc = XDocument.Parse(content);
                    var entries = doc.Descendants("anime").ToList();
                    _importState.Total = entries.Count;
                    _importState.Log($"📄 Found {entries.Count} anime entries in MAL XML export.");

                    for (int i = 0; i < entries.Count; i++)
                    {
                        var animeElem = entries[i];
                        string malIdStr = animeElem.Element("series_animedb_id")?.Value;
                        string title = animeElem.Element("series_title")?.Value?.Trim();
                        string status = animeElem.Element("my_status")?.Value?.Trim();
                        string scoreStr = animeElem.Element("my_score")?.Value?.Trim();

                        if (int.TryParse(malIdStr, out int malId) && !string.IsNullOrEmpty(title))
                        {
                            string category = MalStatusMap.TryGetValue(status ?? "", out string cat) ? cat : (status ?? "Uncategorized");
                            string url = $"https://myanimelist.net/anime/{malId}";
                            
                            double? personalRating = null;
                            if (double.TryParse(scoreStr, out double s) && s != 0)
                            {
                                personalRating = s;
                            }

                            try
                            {
                                bool created = await UpsertAnimeAsync(malId, title, url, category, personalRating);
                                if (created) { added++; _importState.Log($"[{i + 1}/{_importState.Total}] ✅ Added — {title}"); }
                                else { updated++; _importState.Log($"[{i + 1}/{_importState.Total}] 🔄 Updated — {title}"); }
                            }
                            catch (Exception e)
                            {
                                errors++;
                                _importState.Log($"[{i + 1}/{_importState.Total}] ❌ Error for mal_id={malId}: {e.Message}");
                            }
                        }
                        _importState.Current = i + 1;
                    }
                    FinishImport(added, updated, errors);
                    return;
                }
                catch (Exception e)
                {
                    _importState.Log($"❌ XML parse error: {e.Message}");
                    _importState.Status = "error";
                    return;
                }
            }

            // 2. Try JSON
            try
            {
                var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
                var data = JsonSerializer.Deserialize<Dictionary<string, List<JsonElement>>>(content, options);
                
                if (data != null)
                {
                    int total = data.Values.Sum(v => v.Count);
                    _importState.Total = total;
                    _importState.Log($"📄 Found {total} anime entries in JSON format.");

                    int current = 0;
                    foreach (var kvp in data)
                    {
                        string category = kvp.Key;
                        foreach (var item in kvp.Value)
                        {
                            int? malId = item.TryGetProperty("mal_id", out var malIdProp) && malIdProp.ValueKind == JsonValueKind.Number ? malIdProp.GetInt32() : null;
                            string title = item.TryGetProperty("name", out var titleProp) ? titleProp.GetString() : null;
                            string url = item.TryGetProperty("link", out var urlProp) ? urlProp.GetString() : null;

                            if (malId.HasValue && !string.IsNullOrEmpty(title))
                            {
                                try
                                {
                                    bool created = await UpsertAnimeAsync(malId.Value, title, url, category, null);
                                    if (created) { added++; _importState.Log($"[{current + 1}/{total}] ✅ Added — {title}"); }
                                    else { updated++; _importState.Log($"[{current + 1}/{total}] 🔄 Updated — {title}"); }
                                }
                                catch (Exception) { errors++; }
                            }
                            current++;
                            _importState.Current = current;
                        }
                    }
                    FinishImport(added, updated, errors);
                    return;
                }
            }
            catch { }

            // 3. Fallback TXT
            var lines = content.Split(new[] { "\r\n", "\r", "\n" }, StringSplitOptions.None);
            var parsedEntries = new List<(int MalId, string Title, string Url, string Category)>();
            string currentCat = "Uncategorized";

            foreach (var line in lines)
            {
                string l = line.Trim();
                if (string.IsNullOrEmpty(l)) continue;
                if (l.StartsWith("#"))
                {
                    currentCat = l.Substring(1).Trim();
                    continue;
                }
                
                if (l.Contains("|"))
                {
                    var parts = l.Split('|', 2);
                    string title = parts[0].Trim();
                    string url = parts[1].Trim();
                    var match = Regex.Match(url, @"/anime/(\d+)");
                    if (match.Success && int.TryParse(match.Groups[1].Value, out int malId))
                    {
                        parsedEntries.Add((malId, title, url, currentCat));
                    }
                }
            }

            _importState.Total = parsedEntries.Count;
            _importState.Log($"📄 Found {parsedEntries.Count} anime entries in TXT format.");

            for (int i = 0; i < parsedEntries.Count; i++)
            {
                var entry = parsedEntries[i];
                try
                {
                    bool created = await UpsertAnimeAsync(entry.MalId, entry.Title, entry.Url, entry.Category, null);
                    if (created) { added++; _importState.Log($"[{i + 1}/{_importState.Total}] ✅ Added — {entry.Title}"); }
                    else { updated++; _importState.Log($"[{i + 1}/{_importState.Total}] 🔄 Updated — {entry.Title}"); }
                }
                catch (Exception) { errors++; }
                _importState.Current = i + 1;
            }

            FinishImport(added, updated, errors);
        }

        private void FinishImport(int added, int updated, int errors)
        {
            _importState.Log("");
            _importState.Log($"✔️  Import complete! Added: {added} | Updated: {updated} | Errors: {errors}");
            _importState.Status = "done";
        }

        public async Task<bool> UpsertAnimeAsync(int malId, string title, string url, string currentCategory, double? personalRating)
        {
            using var scope = _scopeFactory.CreateScope();
            var db = scope.ServiceProvider.GetRequiredService<AnimeContext>();

            var anime = db.Animes.FirstOrDefault(a => a.MalId == malId);
            bool created = false;

            if (anime == null)
            {
                anime = new Anime
                {
                    MalId = malId,
                    Title = title,
                    Url = url,
                    Category = currentCategory,
                    PersonalRating = personalRating
                };
                db.Animes.Add(anime);
                db.SaveChanges();
                created = true;
            }
            else
            {
                anime.Category = currentCategory;
                if (personalRating.HasValue) anime.PersonalRating = personalRating.Value;
                db.SaveChanges();
                return false;
            }

            try
            {
                await Task.Delay(1500); // Rate limit
                using var response = await _httpClient.GetAsync($"https://api.jikan.moe/v4/anime/{malId}");
                
                string imageUrl = null;
                string synopsis = null;
                
                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    using var doc = JsonDocument.Parse(json);
                    var data = doc.RootElement.GetProperty("data");

                    if (data.TryGetProperty("images", out var images))
                    {
                        if (images.TryGetProperty("webp", out var webp) && webp.TryGetProperty("large_image_url", out var largeImg))
                            imageUrl = largeImg.GetString();
                        else if (images.TryGetProperty("jpg", out var jpg) && jpg.TryGetProperty("large_image_url", out var jpgImg))
                            imageUrl = jpgImg.GetString();
                    }

                    anime.Score = data.TryGetProperty("score", out var scoreProp) && scoreProp.ValueKind == JsonValueKind.Number ? scoreProp.GetDouble() : null;
                    anime.Year = data.TryGetProperty("year", out var yearProp) && yearProp.ValueKind == JsonValueKind.Number ? yearProp.GetInt32() : null;
                    anime.Type = data.TryGetProperty("type", out var typeProp) && typeProp.ValueKind == JsonValueKind.String ? typeProp.GetString() : null;
                    anime.Synopsis = data.TryGetProperty("synopsis", out var synProp) && synProp.ValueKind == JsonValueKind.String ? synProp.GetString() : null;

                    string titleEnglish = data.TryGetProperty("title_english", out var titleEngProp) && titleEngProp.ValueKind == JsonValueKind.String ? titleEngProp.GetString() : null;
                    if (!string.IsNullOrEmpty(titleEnglish))
                    {
                        anime.Title = titleEnglish;
                    }
                }
                else
                {
                    var (kitsuImg, kitsuSyn) = await FetchKitsuFallback(title);
                    imageUrl = kitsuImg;
                    anime.Synopsis = kitsuSyn;
                    if ((int)response.StatusCode == 429)
                    {
                        await Task.Delay(10000);
                    }
                }

                anime.ImageUrl = imageUrl;
                anime.ImageBase64 = await DownloadImageAsBase64(imageUrl);
                db.SaveChanges();
            }
            catch
            {
                // Ignore API fetch errors
            }

            return true;
        }

        private async Task<(string ImageUrl, string Synopsis)> FetchKitsuFallback(string title)
        {
            try
            {
                var request = new HttpRequestMessage(HttpMethod.Get, $"https://kitsu.io/api/edge/anime?filter[text]={Uri.EscapeDataString(title)}&page[limit]=1");
                request.Headers.Add("Accept", "application/vnd.api+json");
                
                var response = await _httpClient.SendAsync(request);
                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    using var doc = JsonDocument.Parse(json);
                    var data = doc.RootElement.GetProperty("data");
                    
                    if (data.GetArrayLength() > 0)
                    {
                        var attrs = data[0].GetProperty("attributes");
                        string imageUrl = null;
                        
                        if (attrs.TryGetProperty("posterImage", out var images))
                        {
                            if (images.TryGetProperty("large", out var large)) imageUrl = large.GetString();
                            else if (images.TryGetProperty("medium", out var med)) imageUrl = med.GetString();
                            else if (images.TryGetProperty("original", out var orig)) imageUrl = orig.GetString();
                        }
                        
                        string synopsis = attrs.TryGetProperty("synopsis", out var syn) ? syn.GetString() : null;
                        return (imageUrl, synopsis);
                    }
                }
            }
            catch { }
            return (null, null);
        }

        public async Task<string> DownloadImageAsBase64(string url)
        {
            if (string.IsNullOrEmpty(url)) return null;
            try
            {
                var response = await _httpClient.GetAsync(url);
                if (response.IsSuccessStatusCode)
                {
                    var bytes = await response.Content.ReadAsByteArrayAsync();
                    var base64 = Convert.ToBase64String(bytes);
                    var contentType = response.Content.Headers.ContentType?.MediaType ?? "image/jpeg";
                    return $"data:{contentType};base64,{base64}";
                }
            }
            catch { }
            return null;
        }
    }
}
